components/User/UserProfile/Integrations.js

```diff
@@ -0,0 +1,390 @@
+        console.log('storage event', e.key, localStorage.getItem(e.key));
+        //this is available
+
+        if (e.key !== NAME || localStorage.getItem(MODI) !== 'addProvider')
+            return false;
+
+        let data = {};
+        try {
+            data = JSON.parse(localStorage.getItem(e.key));
+        } catch (err) {
+            console.log('Error while parsing data', err);
+            this.context.executeAction(updateProviderAction, '');
+            return false;
+        }
+        finally {
+            //delete data
+            localStorage.setItem(NAME, '');
+        }
+
+        //add language before send to service
+        let language = navigator.browserLanguage || navigator.language;
+        if (language.length === 2) {
+            language += '-' + language.toUpperCase();
+        }
+        data.language = language;
+
+        //check data - valid and not empty
+        if ( (data.token.length < 1)
+          || (data.provider.length < 3)
+          || (data.token_creation.length < 22) )
+            //Failure
+        {
+            this.context.executeAction(updateProviderAction, '');
+            return false;
+        }
+
+        if  (data.email.indexOf('@') === -1 || data.email.indexOf('.') === -1 || data.email.length < 5) {
+            //show hint
+            const provider = this.getProviderName();
+            swal({
+                title: 'Error',
+                text: 'The data from ' + provider + ' was incomplete. At least your email should be available for us.',
+                type: 'error',
+                confirmButtonText: 'Confirm',
+                confirmButtonClass: 'negative ui button',
+                buttonsStyling: false
+            }).then().catch();
+            this.context.executeAction(updateProviderAction, '');
+            return false;
+        }
+
+        this.context.executeAction(newSocialData, data);
+
+        //TODO show spinner
+
+
+        this.context.executeAction(addProvider, data);
+
+        return true;
+    }
+
+    getProviderName() {
+        if (this.provider.length < 1)
+            return '';
+        return this.provider.charAt(0).toUpperCase() + this.provider.slice(1);
+    }
+
+    render() {
+        let facebook = false, google = false, github = false;
+        if (this.props.UserProfileStore.user.providers)
+            this.props.UserProfileStore.user.providers.forEach((provider) => {
+                switch (provider) {
+                    case 'facebook':
+                        facebook = true;
+                        break;
+                    case 'google':
+                        google = true;
+                        break;
+                    case 'github':
+                        github = true;
+                        break;
+                }
+            });
+        // console.log('Integrations render()', this.props.UserProfileStore.user.providers);
+
+        let facebook_icon_classes = classNames({
```

Comment:
Depends on your attitude ;) For me, it's easier to read in parameters and I'm confused by repeated functions that do the same thing :P